src/Microsoft.DotNet.Build.CloudTestTasks/UploadToAzure.cs

```diff
@@ -3,94 +3,194 @@
+                "https://{0}.blob.core.windows.net/{1}?restype=container&comp=list", 
+                AccountName, 
+                ContainerName);
+            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(checkListUrl);
+            request.Method = "GET";
+            DateTime dt = DateTime.UtcNow;
+            request.Headers["x-ms-date"] = dt.ToString("R", CultureInfo.InvariantCulture);
+            request.Headers["x-ms-version"] = AzureHelper.StorageApiVersion;
+            request.Headers["Authorization"] = AzureHelper.AuthorizationHeader(
+                AccountName, 
+                AccountKey, 
+                "GET", 
+                dt, 
+                request);
+
+            Log.LogMessage(MessageImportance.Normal, "Sending request to check whether Container blobs exist");
+            HashSet<string> blobsPresent = new HashSet<string>();
+
+            try
             {
-                var relativeBlobPath = item.GetMetadata("RelativeBlobPath");
-                if (string.IsNullOrEmpty(relativeBlobPath))
+                XmlDocument doc;
+                using (WebResponse response = await request.GetResponseAsync())
                 {
-                    Log.LogError(string.Format("Metadata 'RelativeBlobPath' is missing for item '{0}'.", item.ItemSpec));
-                    result = false;
-                    loopState.Stop();
+                    doc = new XmlDocument();
+                    doc.LoadXml(new StreamReader(response.GetResponseStream()).ReadToEnd());
                 }
 
-                CloudBlockBlob blockBlob = container.GetBlockBlobReference(relativeBlobPath);
+                XmlNodeList nodes = doc.DocumentElement.GetElementsByTagName("Blob");
 
-                if (!Overwrite && blockBlob.Exists())
+                foreach (XmlNode node in nodes)
                 {
-                    Log.LogError(string.Format("The blob '{0}' already exists.", blockBlob.Uri));
-                    result = false;
-                    loopState.Stop();
+                    blobsPresent.Add(node["Name"].InnerText);
                 }
-
-                using (var fileStream = File.OpenRead(item.ItemSpec))
+            }
+            catch (WebException e)
+            {
+                Log.LogError(e.Message);
+                Log.LogError(e.Status.ToString());
+                using (Stream stream = e.Response.GetResponseStream())
+                using (StreamReader errorStreamReader = new StreamReader(stream))
                 {
-                    Log.LogMessage(MessageImportance.Low, "Uploading {0} to {1}.", item.ItemSpec, blockBlob.Uri);
-                    blockBlob.UploadFromStream(fileStream);
+                    Log.LogError(errorStreamReader.ReadToEnd());
                 }
-            });
+            }
+
+            Log.LogMessage(MessageImportance.Normal, "Received response to check whether Container blobs exist");
+
+            bool result = true;
+            Parallel.ForEach(
```

Comment:
SInce Upload is Async now, you don't actually need to parallel this, because it doesn't block on web requests anymore.

Factor out the body of the lamba (which is good practice anyway, so you can keep track of what values are going to be referenced by multiple callers and which one).

Then ditch the Parallel, and just do:
await Task.WhenAll(Items.Select(item => MyRefactoredMethod(item))

Parallel usually brings a lot more complications to the party than is necessary, and since we aren't actually doing a lot of parallel CPU bound work here, those complications probably outweigh the benefits.